### ISSUE TITLE
Another assertion failure with invalid input

### ISSUE THREAD

**Issue by psychon** [team-charls/charls#37]
Since #36 got a little crowded and hard to follow, I'll report this as a separate issue. Feel free to tell me how you prefer to get these kind of reports (I guess that there will be more to come, sorry).

Anyway, the following program dies with an assertion failure: `/tmp/charls/src/decoderstrategy.h:129: void DecoderStrategy::MakeValid(): Assertion 'static_cast<size_t>(_validBits) <=bufType_bit_count - 8' failed.`
```c
#include "src/charls.h"

const unsigned char input[] = {
  0xff, 0xd8, 0xff, 0xf7, 0x00, 0x11, 0x08, 0x00, 0x01, 0x00, 0xfd, 0x02,
  0x01, 0x11, 0x00, 0x02, 0x11, 0x00, 0x03, 0x11, 0x00, 0xff, 0xda, 0x00,
  0x08, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x80, 0x10, 0xd9
};

int main()
{
	char result[1024 * 1024];
	JpegLsDecode(result, sizeof(result), input, sizeof(input), nullptr, nullptr);
	return 0;
}
```
The problem seems to be... I don't know. `ReadHighbits()` has `Peek0Bits()` returning `-1` and so calls `Skip(15)`. However, `_validBits` is less than 15 at this point, so becomes negative, which causes the above assertion failure when `MakeValid()` is called the next time (casting a negative value to unsigned produces something huge and... well). So the "real error" happened way earlier than the assertion failure.

I'm not quite sure who is to blame here. `ReadHighbits()` assumes that `_validBits >= 15` (otherwise `Skip(15)` is malformed), but this assumption is supposedly guaranteed by `Peek0Bits()` calling `MakeValid()` if there are less than 16 bits available. However, `MakeValid()` doesn't seem to do anything since the end of the input was already reached (`_position >= _endPosition`). So, $SOMETHING has to check for EOF, but I wonder how general this check has to be (in `ReadHighbits()`? In `Skip()`?).

**Comment by hammeramr** [comment]
Hey @psychon any luck making it work ? 

I have also been dealing with one assertion failure after another. Please let me know if you have had any luck encoding and decoding with any JPEG - LS library. 

Also I have been tinkering around with [https://github.com/thorfdbg/libjpeg](libjpeg) which works well in the command line but I am having a hard time using it as an API and making it use buffers instead of writing and reading from disk.

Thanks,
- Amr

**Comment by psychon** [comment]
@hammeramr I had no problems with charls when I wanted to use for encoding and decoding images. This problem was only found when specifically looking for a problem: I ran american fuzzy lop against CharLS. AFL specifically tries to find malformed inputs that make a program crash.

So, I did not run into this assertion failure accidentally and it does not mean that CharLS is unusable for me.

**Comment by hammeramr** [comment]
@psychon thanks for the quick reply, I removed the failure by picking an interleave mode. However now I am getting: err: Color transformation 32762 is not supported.: CharLS error


Do you have an example where you encode/decode a colour image ? If you do may I have a look at it please ?

Than you for your time

**Comment by psychon** [comment]
Decode: http://git.dcmtk.org/?p=dcmtk.git;a=blob;f=dcmjpls/libsrc/djcodecd.cc;h=3f0d85e1df2d479dd1f1a3c9a60e8d21a4d3f073;hb=HEAD
Encode: http://git.dcmtk.org/?p=dcmtk.git;a=blob;f=dcmjpls/libsrc/djcodece.cc;h=d13da73cde3f6d7972e31988780b6688eea84fc5;hb=HEAD

However, I guess you would find the code in this repository more useful. For example, `TestRoundTrip` in `test/util.cpp` can encode and decode color images (`components = 3`).

**Comment by hammeramr** [comment]
Thank you very much :+1:

**Comment by vbaderks** [comment]
This issue was closed by #53
